service/src/main/java/bio/terra/pipelines/notifications/NotificationService.java

```diff
@@ -0,0 +1,175 @@
+    // in the quotas table
+    UserQuota userQuota =
+        quotasService.getOrCreateQuotaForUserAndPipeline(userId, pipeline.getName());
+    String quotaRemaining = String.valueOf(userQuota.getQuota() - userQuota.getQuotaConsumed());
+
+    return new BaseTeaspoonsJobNotification(
+        pipelineRun.getUserId(),
+        pipelineDisplayName,
+        pipelineRun.getJobId().toString(),
+        formatInstantToReadableString(pipelineRun.getCreated()),
+        formatInstantToReadableString(pipelineRun.getUpdated()),
+        quotaRemaining,
+        pipelineRun.getDescription());
+  }
+
+  /**
+   * Format an Instant as a date time string in UTC using the RFC-1123 date-time formatter, such as
+   * 'Tue, 3 Jun 2008 11:05:30 GMT'.
+   *
+   * @param dateTime the Instant to format
+   * @return the formatted date time string
+   */
+  public String formatInstantToReadableString(Instant dateTime) {
+    return dateTime.atZone(ZoneId.of("UTC")).format(DateTimeFormatter.RFC_1123_DATE_TIME);
+  }
+
+  /**
+   * Create a notification object for a failed job.
+   *
+   * @param jobId the job id
+   * @param userId the user id
+   * @param context the flight context
+   * @return the notification object
+   */
+  public TeaspoonsJobFailedNotification createTeaspoonsJobFailedNotification(
+      UUID jobId, String userId, FlightContext context) {
+    // get exception
+    Optional<Exception> exception = context.getResult().getException();
+    String errorMessage;
+    if (exception.isPresent()) {
+      ApiErrorReport errorReport =
+          buildApiErrorReport(exception.get()); // use same logic that the status endpoint uses
+      errorMessage = errorReport.getMessage();
+    } else {
+      logger.error(
+          "No exception found in flight result for flight {} with status {}",
+          context.getFlightId(),
+          context.getFlightStatus());
+      errorMessage = "Unknown error";
+    }
+
+    return new TeaspoonsJobFailedNotification(
+        createBaseTeaspoonsJobNotification(jobId, userId), errorMessage);
+  }
+
+  /**
+   * Create a notification object for a successful job.
+   *
+   * @param jobId the job id
+   * @param userId the user id
+   * @return the notification object
+   */
+  public TeaspoonsJobSucceededNotification createTeaspoonsJobSucceededNotification(
+      UUID jobId, String userId) {
+    String quotaConsumedByJob =
+        pipelineRunsService.getPipelineRun(jobId, userId).getQuotaConsumed().toString();
+    return new TeaspoonsJobSucceededNotification(
+        createBaseTeaspoonsJobNotification(jobId, userId), quotaConsumedByJob);
+  }
+
+  /**
+   * Configure and send a notification that a job has succeeded.
+   *
+   * @param jobId the job id
+   * @param userId the user id
+   */
+  public void configureAndSendPipelineRunSucceededNotification(UUID jobId, String userId) {
+    try {
+      pubsubService.publishMessage(
```

Comment:
what information does thurloe use in the message to figure out what template to use and what values to put where?

Reply:
the `notificationType` string is recognized by Thurloe and converted into the SendGrid template ID (mapping defined in terra-helmfile: https://github.com/broadinstitute/terra-helmfile/blob/master/charts/thurloe/values.yaml#L47)